File: 0074_search_a_2d_matrix.swift
// https://leetcode.com/problems/search-a-2d-matrix/

class Solution {
	func searchMatrix(_ matrix: [[Int]], _ target: Int) -> Bool {
		if target < matrix.first!.first! || target > matrix.last!.last! {
			return false
		}

		var l = 0
		var r = matrix.count - 1
		while l <= r {
			let mid = (l + r) / 2
			let midVal = matrix[mid]

			if target < midVal.first! {
				r =	mid - 1
			} else if target > midVal.last! {
				l = mid + 1
			} else {
				return binary2(array: midVal, target: target)
			}
		}

		return false
	}

	private func binary2(array: [Int], target: Int) -> Bool {
		var l = 0
		var r = array.count - 1

		while l <= r {
			let mid = (l + r) / 2
			let midVal = array[mid]

			if target == midVal {
				return true
			} else if target < midVal {
				r = mid - 1
			} else {
				l = mid + 1
			}
		}

		return false
	}
}

File: 0074_search_a_2d_matrix.kt
// https://leetcode.com/problems/search-a-2d-matrix/description/

class Solution {
    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        val colSize = matrix[0].size
        var l = 0
        var r = matrix.size - 1
        while (l <= r) {
            val mid = (l + r) / 2
            if (target < matrix[mid][0]) {
                r = mid - 1
            } else if (target > matrix[mid][colSize - 1]) {
                l = mid + 1
            } else {
                return searchRow(matrix[mid], target)
            }
        }
        return false
    }

    private fun searchRow(row: IntArray, target: Int): Boolean {
        var l = 0
        var r = row.size - 1
        while (l <= r) {
            val mid = (l + r) / 2
            if (target == row[mid]) {
                return true
            } else if (target < row[mid]) {
                r = mid - 1
            } else {
                l = mid + 1
            }
        }
        return false
    }
}
